### PR TITLE
Feat:Add Gateway Plugin Support 

### LINKED ISSUE #651: Add Gateway Plugin Support
**What would you like to be added**:
Extend the gateway plugins supported by Kurator's release feature, initially including NGINX and Kuma.
**Why is this needed**:
Simplify user configuration and provide more options for traffic routing.

### PR DESCRIPTION
**Pull request by Gidi233**
**What type of PR is this?**
/kind api-change
/kind feature

**What this PR does / why we need it**:
Users can choose kuma and nginx in the configuration of rollout and fleet flagger.
Specifically, you can set the chart and value of the helm deployment gateway in the fleet configuration.
You can specify the host of Ingress (required) and the protocol used by Kuma (default is http) in the rollout configuration.
Ingress can automatically complete the deployment of canary through application.
Kuma needs to manually install Prometheus and mesh in the target cluster.

**Which issue(s) this PR fixes**:
Fixes #651

**Does this PR introduce a user-facing change?**:
For application:
```yaml
    ......
      rollout:
        trafficRoutingProvider: nginx
        rolloutPolicy:
          trafficRouting:
            ......
            host: "app.example.com"
---
    ......
      rollout:
        trafficRoutingProvider: kuma
        rolloutPolicy:
          trafficRouting:
            ......
            protocol: http
```
For Fleet
```yaml
  ......
  plugin:
    flagger:
      trafficRoutingProvider: nginx
      config:
        chart: 
        extraArgs: 
```

### CODE REVIEW

**Review by LiZhenCheng9527 — COMMENTED** [review]
Your PR is too large.
In fact, it can be split into multiple PRs such as API changes, fleet plugin installation implementation, feature implementation, etc. 
It is convenient for development and review.

File: pkg/apis/apps/v1alpha1/types.go [hunk 1]
```diff
@@ -239,6 +239,39 @@ type TrafficRoutingConfig struct {
...
+	//                 port:
+	//                   number: application.syncPolicies.rollout.port
+	// ```
+	// +optional
+	Host string `json:"host,omitempty"`
+
+	// for kuma
+	// Defaults to http
+	// +optional
+	Protocol string `json:"protocol,omitempty"`
```
[the remark this answers is not in the document]
Comment (Gidi233) [reply pkg/apis/apps/v1alpha1/types.go:273]: yes

File: pkg/apis/apps/v1alpha1/types.go [hunk 2]
```diff
@@ -239,6 +239,39 @@ type TrafficRoutingConfig struct {
...
+	//           - pathType: Prefix
+	//             path: "/"
+	//             backend:
+	//               service:
+	//                 name: application.syncPolicies.rollout.ServiceName
+	//                 port:
+	//                   number: application.syncPolicies.rollout.port
+	// ```
+	// +optional
+	Host string `json:"host,omitempty"`
```
[the remark this answers is not in the document]
Comment (Gidi233) [reply pkg/apis/apps/v1alpha1/types.go:268]: I mean, you only need to fill in the host field as the host of the ingress. The other default items will be hard-coded in the code as configured above.

File: pkg/fleet-manager/application/rollout_helper.go [hunk 3]
```diff
@@ -397,14 +445,88 @@ func deleteResourceCreatedByKurator(ctx context.Context, namespaceName types.Nam
...
+		}
+	} else {
+		// verify if the ingress were created by kurator
+		annotations := ingress.GetAnnotations()
+		if _, exist := annotations[RolloutIdentifier]; exist {
+			labels := ingress.GetLabels()
+			if set := sets.New(strings.Split(labels[ingressLabelKey], ",")...); set.Contains(rollout.ServiceName) {
+				if set.Len() == 1 {
+					if deleteErr := kubeClient.Delete(ctx, ingress); deleteErr != nil && !apierrors.IsNotFound(deleteErr) {
+						return errors.Wrapf(deleteErr, "failed to Delete ingress %s in %s", namespaceName.Name, namespaceName.Namespace)
```
Comment (LiZhenCheng9527) [inline pkg/fleet-manager/application/rollout_helper.go:466]: nit:
```suggestion
						return errors.Wrapf(deleteErr, "failed to Delete ingress %s in %s", rollout.Workload.Namespace, ingress.Name)
```

File: pkg/fleet-manager/application/rollout_helper.go [hunk 4]
```diff
@@ -397,14 +445,88 @@ func deleteResourceCreatedByKurator(ctx context.Context, namespaceName types.Nam
...
+					for _, rule := range ingress.Spec.Rules {
+						if rule.Host != rollout.RolloutPolicy.TrafficRouting.Host {
+							newRules = append(newRules, rule)
+						}
+					}
+					ingress.Spec.Rules = newRules
+					labels[ingressLabelKey] = strings.Join(set.Delete(rollout.ServiceName).UnsortedList(), ",")
+					ingress.SetLabels(labels)
+					if err := kubeClient.Update(ctx, ingress); err != nil {
+						return errors.Wrapf(err, "failed to Update ingress %s in %s", namespaceName.Name, namespaceName.Namespace)
```
Comment (LiZhenCheng9527) [inline pkg/fleet-manager/application/rollout_helper.go:479]: ditto

File: pkg/fleet-manager/application/rollout_helper.go [hunk 5]
```diff
@@ -397,14 +445,88 @@ func deleteResourceCreatedByKurator(ctx context.Context, namespaceName types.Nam
...
+					newRules := make([]ingressv1.IngressRule, 0)
+					for _, rule := range ingress.Spec.Rules {
+						if rule.Host != rollout.RolloutPolicy.TrafficRouting.Host {
+							newRules = append(newRules, rule)
+						}
+					}
+					ingress.Spec.Rules = newRules
+					labels[ingressLabelKey] = strings.Join(set.Delete(rollout.ServiceName).UnsortedList(), ",")
+					ingress.SetLabels(labels)
+					if err := kubeClient.Update(ctx, ingress); err != nil {
```
Comment (LiZhenCheng9527) [inline pkg/fleet-manager/application/rollout_helper.go:478]: Why put `update` and `delete` in one function?
Separate these two functions

File: pkg/fleet-manager/application/rollout_helper.go [hunk 6]
```diff
@@ -397,14 +445,88 @@ func deleteResourceCreatedByKurator(ctx context.Context, namespaceName types.Nam
...
+					}
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// create/update ingress configuration
+func renderIngress(ingress *ingressv1.Ingress, rollout *applicationapi.RolloutConfig) {
```
Comment (LiZhenCheng9527) [inline pkg/fleet-manager/application/rollout_helper.go:489]: Is this only used by nginx?
Maybe the function name can be replaced to `renderNginxIngress`

File: pkg/fleet-manager/application/rollout_helper.go [hunk 7]
```diff
@@ -569,6 +705,19 @@ func addLabels(obj client.Object, key, value string) client.Object {
 	return obj
 }
 
+func addAnnotations(obj client.Object, ann map[string]string) client.Object {
+	annotations := obj.GetAnnotations()
+	if annotations == nil {
+		obj.SetAnnotations(ann)
+	} else {
+		for k, v := range ann {
+			annotations[k] = v
```
Comment (LiZhenCheng9527) [inline pkg/fleet-manager/application/rollout_helper.go:714]: This may change the original annotations, which need to be evaluated to see if there is any .

**Comment by LiZhenCheng9527** [comment]
almost LGTM

File: pkg/apis/apps/v1alpha1/types.go [hunk 8]
```diff
@@ -239,6 +239,42 @@ type TrafficRoutingConfig struct {
 	// +optional
 	CorsPolicy *istiov1alpha3.CorsPolicy `json:"corsPolicy,omitempty"`
 
+	// Host defines the domain name you specify.
```
[the remark this answers is not in the document]
Comment (Gidi233) [reply pkg/apis/apps/v1alpha1/types.go:242]: yes.

File: pkg/fleet-manager/application/rollout_helper.go (hunk 6 above)
Comment (Gidi233) [reply pkg/fleet-manager/application/rollout_helper.go:489]: ok

File: pkg/fleet-manager/application/rollout_helper.go (hunk 7 above)
Comment (Gidi233) [reply pkg/fleet-manager/application/rollout_helper.go:714]: checked

File: pkg/fleet-manager/application/rollout_helper.go (hunk 5 above)
Comment (Gidi233) [reply pkg/fleet-manager/application/rollout_helper.go:478]: fixed

File: pkg/fleet-manager/application/rollout_helper.go [hunk 9]
```diff
@@ -397,14 +445,88 @@ func deleteResourceCreatedByKurator(ctx context.Context, namespaceName types.Nam
...
+			return errors.Wrapf(err, "failed to get ingress %s in %s", namespaceName.Name, namespaceName.Namespace)
+		}
+	} else {
+		// verify if the ingress were created by kurator
+		annotations := ingress.GetAnnotations()
+		if _, exist := annotations[RolloutIdentifier]; exist {
+			labels := ingress.GetLabels()
+			if set := sets.New(strings.Split(labels[ingressLabelKey], ",")...); set.Contains(rollout.ServiceName) {
+				if set.Len() == 1 {
+					if deleteErr := kubeClient.Delete(ctx, ingress); deleteErr != nil && !apierrors.IsNotFound(deleteErr) {
```
[the remark this answers is not in the document]
Comment (Gidi233) [reply pkg/fleet-manager/application/rollout_helper.go:465]: Okay, I'll delete it.

**Comment by LiZhenCheng9527** [comment]
Please fix lint error

**Review by hzxuzhonghu — APPROVED** [review]
Thanks

**Pull request merged**